### PR TITLE
Many improvements

### PR DESCRIPTION
**Pull request by hpbuniat**
Sorry for not making this single PR's but i was in some kind of flow ;)

In a nutshell:
- added new codes to identify a temporary ban
- fixed the futbin update-time parser to work with a day-range (e.g. 4 days ago)
- moved the number of auctions to try to a configuration (via .env-file)
- added price-information to the investor-report
- the investor-loop now checks the optimal-sell-price only once to reduce api-calls (was: twice)
- there's a new resource "invest/low-players-relist" to re-list all expired auctions and adjust the price if necessary (the new price will be checked agains the bought-price to prevent loss, but you have to resolve this manually)
- added prp-query parameter to the initial futbin call, to query only those players that have a sort of reliable price-information on futbin (see https://www.futbin.com/prp)
- futbin-sell-prices are also double checked with updated-time and prp
- the market-price-calculation now ignores the default "10000"-prices and skips if there are too many "invalid" auctions (e.g. default-prices) or too few samples for the price

### CODE REVIEW

File: lib/invest/invest-app.ts [hunk 1]
```diff
@@ -19,24 +19,29 @@ investApp.get("/players", async (req, res) => {
...
 investApp.get("/low-players", async (req, res) => {
-  const { maxTargetPool } = req.query;
-  const { budget, min, max } = getBudgetMinMax(req.query);
+  const {maxTargetPool} = req.query;
+  const {budget, min, max} = getBudgetMinMax(req.query);
   if (!budgetMinMaxQueryCheck(res, budget, min, max)) return;
 
   res.send(
-    investService.startLowPlayerInvvest({ budget, min, max, maxTargetPool })
+    investService.startLowPlayerInvvest({budget, min, max, maxTargetPool})
```
Comment (joelmenezes) [inline lib/invest/invest-app.ts]: nit: `startLowPlayerInvvest` should be `startLowPlayerInvest`

File: .env.local [hunk 2]
```diff
@@ -1,6 +1,8 @@
 # Fut api
-FUTBOT_FUT_REQUESTS_PER_SEC=.7
+FUTBOT_FUT_MAX_AUCTION_TRY_PER_PLAYER=3
+FUTBOT_FUT_MINIMUM_AUCTION_SAMPLES=3
+FUTBOT_FUT_REQUESTS_PER_SEC=.5
```
Comment (dogancana) [inline .env.local]: I'm not sure if it's wise to have these values to be configurable. The values would effect how efficient a job executes. And it's not really clear to someone who is not maintaining the project. 
What was the reason for you to move them here?

File: lib/api/fut-api/api.ts [hunk 3]
```diff
@@ -63,10 +63,16 @@ futApi.interceptors.response.use(
       Job.stopAllJobs();
     }
 
-    if (status === 429) {
+    if ([512, 521].indexOf(status) > -1) {
```
Comment (dogancana) [inline lib/api/fut-api/api.ts]: can you also add 458 here. It's 'Verification required' error

File: lib/invest/invest-service.ts [hunk 4]
```diff
@@ -40,23 +46,66 @@ export namespace investService {
     return playerInfos;
   }
 
+  export async function reListLowPlayer(): Promise<string[]> {
```
Comment (dogancana) [inline lib/invest/invest-service.ts]: I think this is more of a trade-bot job than invest job. Let's move it there

File: lib/invest/jobs/low-player-investor.ts [hunk 5]
```diff
@@ -87,26 +96,29 @@ export class LowPlayerInvestor extends Job {
...
-            assetId: target.assetId
-          });
-          this.spent += lowest.buyNowPrice;
-          this.budget -= lowest.buyNowPrice;
 
           const sellTarget = await fut.waitAndGetPurchasedItem(
             target.resourceId
           );
           if (sellTarget) {
+            this.boughtPlayers.push({
```
Comment (dogancana) [inline lib/invest/jobs/low-player-investor.ts:112]: I liked this change :)

File: lib/player/player-service.ts [hunk 6]
```diff
@@ -13,49 +13,74 @@ export namespace playerService {
...
       minStartingBid: Number.MAX_VALUE,
       averageBuyNow: 0,
       averageStartingBid: 0,
-      samplecount: 0
+      sampleCount: 0
     };
 
     auctions.forEach(a => {
-      if (a.itemData.resourceId != resourceId) return;
+      if (a.itemData.resourceId != resourceId) {
```
Comment (dogancana) [inline lib/player/player-service.ts:33]: I don't remember why I put this check here. Did you fall into this case while testing?

File: lib/player/player-service.ts [hunk 7]
```diff
@@ -13,49 +13,74 @@ export namespace playerService {
...
+        invalidSamples++;
+        return;
+      }
+
+      // ignore the "default" 10000 buy-now price
+      if (a.buyNowPrice === 10000) {
```
Comment (dogancana) [inline lib/player/player-service.ts]: I think we should change this change to a.buyNowPrice === a.itemData.maxPrice
I don't know if it's called maxPrice but there was a field indicating this default buy now price value

File: lib/player/player-service.ts [hunk 8]
```diff
@@ -13,49 +13,74 @@ export namespace playerService {
...
 
-    price.minBuyNow =
-      price.minBuyNow === Number.MAX_VALUE ? 0 : price.minBuyNow;
+    // prevent division by zero
+    if (price.sampleCount > 0) {
+      price.averageBuyNow /= price.sampleCount;
+      price.averageStartingBid /= price.sampleCount;
+    }
 
+    // if there are more invalid samples than valid ones, invalidate the player
```
Comment (dogancana) [inline lib/player/player-service.ts]: I don't this this is correct. There will be cases where a player will be listed 10 times with buy now price of 10000 and buy now price of 1000 for 5 times. This doesn't make player's samples invalid. His price should return as 1000

File: lib/trader/trade-utils.ts [hunk 9]
```diff
@@ -27,22 +27,29 @@ export interface SellPrice {
...
 
-  if (price.updatedMinsAgo > 60) {
+  // ignore old prices
+  if (price.updatedMinsAgo === -1 || price.updatedMinsAgo > 60) {
     return null;
   }
 
-  let referencePrice = prices[0] * 1.05;
+  // ignore low price-range percentage (https://www.futbin.com/prp)
+  if (price.prp < 25) {
```
Comment (dogancana) [inline lib/trader/trade-utils.ts]: nice check

**Review by dogancana — CHANGES_REQUESTED** [review]
Looks good in general
Thanks for improvements!
Let's discuss over my comments and I can also fix some of them later today

File: lib/api/fut-api/api.ts (hunk 3 above)
Comment (hpbuniat) [reply lib/api/fut-api/api.ts]: Good catch :)

File: lib/player/player-service.ts (hunk 7 above)
Comment (hpbuniat) [reply lib/player/player-service.ts]: It's `marketDataMaxPrice`. Changed it 👍

File: lib/player/player-service.ts (hunk 6 above)
Comment (hpbuniat) [reply lib/player/player-service.ts:33]: No, i was just adding braces ;)

File: lib/invest/invest-service.ts (hunk 4 above)
Comment (hpbuniat) [reply lib/invest/invest-service.ts]: Check 👍

File: lib/player/player-service.ts (hunk 8 above)
Comment (hpbuniat) [reply lib/player/player-service.ts]: Sure thing. I guess the other safety checks are enough to ensure a sure price-assumption.
Removed it 👍

**Comment by hpbuniat** [comment]
I added a few things next to the comments. In a nutshell:
- bid-queries are bypassing the queue, to act faster
- if there's an auth-error, etc. the queue is cleared
- some safety assumptions are now a bit more loose again, to reduce requests agains the FUT-api

I will resolve the conflicts later this week ;)

File: .env.local (hunk 2 above)
Comment (hpbuniat) [reply .env.local]: Maybe a short description can help with that. I'm always trying to reduce "magic numbers" - especially when they are used more than once throughout the whole project :).

Comment (dogancana) [reply .env.local]: Sounds good. Defaults should prevent any issues to some point

**Pull request merged**